src/storage/accessors/S3DataAccessor.ts

```diff
@@ -0,0 +1,274 @@
+/* eslint-disable @typescript-eslint/naming-convention -- AWS S3 API uses capitalized keys */
+import { Readable } from 'stream';
+import type { CopyObjectCommandOutput, DeleteObjectCommandOutput, GetObjectCommandOutput, HeadObjectCommandOutput,
+  ListObjectsV2CommandOutput, PutObjectCommandOutput, ServiceInputTypes, ServiceOutputTypes } from '@aws-sdk/client-s3';
+import { MetadataDirective, S3ServiceException, S3 } from '@aws-sdk/client-s3';
+import { DataFactory } from 'n3';
+import type { Representation } from '../../http/representation/Representation.js';
+import { RepresentationMetadata } from '../../http/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../../http/representation/ResourceIdentifier';
+import { getLoggerFor } from '../../logging/LogUtil';
+import { InternalServerError } from '../../util/errors/InternalServerError';
+import { NotFoundHttpError } from '../../util/errors/NotFoundHttpError';
+import { UnsupportedMediaTypeHttpError } from '../../util/errors/UnsupportedMediaTypeHttpError';
+import type { Guarded } from '../../util/GuardedStream';
+import { guardStream } from '../../util/GuardedStream';
+import { ensureLeadingSlash, ensureTrailingSlash, isContainerIdentifier, joinUrl } from '../../util/PathUtil';
+import { addResourceMetadata } from '../../util/ResourceUtil';
+import { DC, HH, LDP, MA, RDF, XSD } from '../../util/Vocabularies';
+import type { DataAccessor } from './DataAccessor';
+
+/** Metadata useable with the S3 SDK */
+interface S3Metadata {
+  Metadata?: Record<string, string>;
+  ContentType?: string;
+  ContentLength?: number;
+  LastModified?: Date;
+}
+
+/** The metadata header prefix. */
+const s3prefix = 'x-amz-meta-';
+
+/**
+ * Transforms RepresentationMetadata to S3Metadata, quoting literals.
+ * Note: header prefixes are added by the S3 SDK.
+ */
+function encodeMetadata(metadata: RepresentationMetadata): S3Metadata {
+  const { contentType: ContentType, contentLength: ContentLength } = metadata;
+
+  metadata.removeAll(MA.terms.format);
+  metadata.removeAll(DC.terms.modified);
+  metadata.removeAll(HH.terms['content-length']);
+  metadata.remove(RDF.terms.type, LDP.terms.Resource);
+  metadata.remove(RDF.terms.type, LDP.terms.Container);
+  metadata.remove(RDF.terms.type, LDP.terms.BasicContainer);
+
+  const Metadata = Object.fromEntries(metadata.quads().map(
+    ({ predicate, object }): [string, string] => [
+      encodeURIComponent(predicate.value),
+      object.termType === 'Literal' ? `"${object.value}"` : object.value,
+    ],
```

Comment:
This is going to lose data if a predicate has multiple objects. Would also recommend just using something like the `rdf-string` library to prevent any potential edge cases. `rdf-string` can then also be used when converting back to terms.

Reply:
Yes, I was aware of that. I'll look into it to see if we can use S3 metadata with lists as values.